src/traefik.py

```diff
@@ -0,0 +1,600 @@
+                        "default": {
+                            # When the external hostname is a bare IP, traefik cannot match a domain,
+                            # so we must set the default cert for the TLS handshake to succeed.
+                            "defaultCertificate": {
+                                "certFile": SERVER_CERT_PATH,
+                                "keyFile": SERVER_KEY_PATH,
+                            },
+                        },
+                    },
+                }
+            }
+        )
+        self._container.push(DYNAMIC_CERTS_PATH, config, make_dirs=True)
+
+    def configure(self):
+        """Configure static and tls."""
+        # Ensure the required basic configurations and folders exist
+        tcp_entrypoints = self._tcp_entrypoints
+        self._update_static_configuration(tcp_entrypoints)
+        self._setup_dynamic_config_folder()
+
+        if self._tls_enabled:
+            self._update_tls_configuration()
+
+    def update_cert_configuration(
+        self, cert: Optional[str], key: Optional[str], ca: Optional[str]
+    ):
+        """Update the server cert, ca, and key configuration files."""
+        if cert:
+            self._container.push(SERVER_CERT_PATH, cert, make_dirs=True)
+        else:
+            self._container.remove_path(SERVER_CERT_PATH, recursive=True)
+
+        if key:
+            self._container.push(SERVER_KEY_PATH, key, make_dirs=True)
+        else:
+            self._container.remove_path(SERVER_KEY_PATH, recursive=True)
+
+        if ca:
+            self._container.push(CA_CERT_PATH, ca, make_dirs=True)
+        else:
+            self._container.remove_path(CA_CERT_PATH, recursive=True)
+
+        self.update_ca_certs()
+
+    def add_ca(self, ca: str, uid: Union[str, int]):
+        """Add a ca."""
+        ca_path = RECV_CA_TEMPLATE.substitute(rel_id=str(uid))
+        self._container.push(ca_path, ca, make_dirs=True)
+
+    def remove_ca(self, uid: Union[str, int]):
+        """Remove a ca.
+
+        BEWARE of potential race conditions.
+        Traefik watches the dynamic config dir and reloads automatically on change.
+        So make sure any traefik config depending on the certificates being there is updated
+        **before** the certs are removed, otherwise you might have some downtime.
+        """
+        ca_path = RECV_CA_TEMPLATE.substitute(rel_id=str(uid))
+        self._container.remove_path(ca_path)
+
+    def update_ca_certs(self):
+        """Update ca certificates and restart traefik."""
+        self._container.exec(["update-ca-certificates", "--fresh"]).wait()
+
+        # Must restart traefik after refreshing certs, otherwise:
+        # - newly added certs will not be loaded and traefik will keep erroring-out with "signed by
+        #   unknown authority".
+        # - old certs will be kept active.
+        self.restart()
+
+    def _update_static_configuration(self, tcp_entrypoints: Dict[str, int]):
```

Comment:
Isn't `tcp_entrypoints` a constructor arg?

Reply:
ha, you're right. forgot to lift that